
[![License](http://img.shields.io/:license-mit-blue.svg)](http://doge.mit-license.org)

# gene_flow_linkage

This repository contains all the scripts and processed data files used to perform the data processing and analyses described in:

**Gene flow favors linkage between adaptive alleles in a globally distributed species**

*Kieran Samuk, Greg Owens, Diana Rennison, Kira Delmore, Sara Miller and Dolph Schluter.*

If you have trouble getting the code here to function, or have questions about the dataset, feel free to post an issue. Otherwise, contact Kieran Samuk at samuk@zoology.ubc.ca.

### Requirements

- **R version 3.2.2**. Certain dependencies will only work with this version of R. If you are using Rstudio, you can toggle the current version of R in Global Options.
- **[Rstudio](https://www.rstudio.com/) (any version that supports packrat)**. This is streamlines the use of [packrat](https://rstudio.github.io/packrat/).

### How to use this repo

#### I want to look at your code!

All code can be free browsed online in this repository.

#### I want to look at and/or use your data!

All raw data used in this project are available in the following locations:

1. DNA sequence - on the NCBI SRA here and here (link forthcoming)
2. Divergence Data (SNP-level) - on Dryad here (link forthcoming).
3. Divergence Data (window-level)- in the packrat bundle [here](https://www.dropbox.com/s/6csk93s65u3q1px/gene_flow_linkage_bundle_V1.tar.gz?dl=0)

#### I want to reproduce your analysis, with your data or possibly mine!

Reccomended method:

1. Download a zip of this repo, including data and a bundle of package dependencies [HERE](https://www.dropbox.com/s/6csk93s65u3q1px/gene_flow_linkage_bundle_V1.tar.gz?dl=0)
2. Open the provided .Rproj file
3. packrat will initialize the provided dependency library.
4. Run/browse code etc.

When using the packrat bundle, the following top-level scripts will 'as-is':

* 04_analyze_75k_permute_S2.R
* 05_analyze_clustering_output_S3.R
* 06_plot_figure_1.R
* 07_plot_figures_2.R
* 08_plot_figures_S4_S5.R
* 09_plot_figure_S1.R
* 10_prepare_supplemental_data_files.R
* 11_run_all_analyze_plot_scripts.R

The remaining below require precursors to function (these will be made available on Dryad, accession number to follow):

* 01_process_75k_fit_models.R
* 02_process_snp_clustering_fst.R
* 03_process_clustering_output.R

#### I'm not sure what I want to do!

That's OK. May I suggest some [festive music](https://player.spotify.com/album/7vHHpcOnYCjZhYq0caS1qE), in that case? 

### Repository contents, by folder

#### analysis_ready
Processed divergence data for populations used in the project. This is the 'raw'(ish) data used for the analysis. These files are generated by scripts 1-3 (the 'process' scripts). 

#### ev_prep_scripts
The scripts used to generate the explanatory variables ('evs;) used in our regressions. Recombination rate, mutation rate, and genet density calculations are found here.

#### evs
The explanatory variables used in the regressions (recombination, mutation, gene density). These are generated by the ev_prep_scripts

#### figures
These are the figures from the manuscript. The 'plot' scripts output figures to the `figures/raw` folder. The figures in the root `figures` folder are the good copies of the figures, post formating in Inkscape.

#### meta_data
Miscellaneous metadata, including population locations, overall FST estimates, p-values, etc.

#### process_raw_data
The scripts used to do all analyses upstream of the ones outlined here. These scripts move from raw .fastq data (e.g. from a database), to population genetic summary stastistics. A separate readme is provided for this section.

#### shared_functions
The library of functions used by the top-level scripts.
